File: P6_OPM.c
#include <stdio.h>
#include <string.h>

struct prod {
    char lhs[50];
    char rhs[50];
} p[30];
int first[20][20], firstp[20][20], firsts[20][20], last[20][20], lastp[20][20], lasts[20][20], equals[20][20], lt[20][20], gt[20][20];
char opm[20][20];
int tot, i, j, k, cnt = 0, flag = 0, sum = 0;
char sym[30], ans = 'y', ch = '\0';

void display(int mat[20][20]) {
    for (i = 0; i < strlen(sym); i++) {
        printf(" %c", sym[i]);
    }
    printf("\n");
    for (i = 0; i < strlen(sym); i++) {
        printf("%c ", sym[i]);
        for (j = 0; j < strlen(sym); j++) {
            printf("%d ", mat[i][j]);
        }
        printf("\n");
    }
}

int main() {
    tot = 0;

    while (ans == 'y') {
        printf("Saail Chavan KFPMSCCS016");
        printf("\n\n Enter production (LHS RHS) : ");
        scanf("%s %s", p[tot].lhs, p[tot].rhs);
        tot++;
        printf("\n continue? (y/n) : ");
        ans = getchar();
        // Clear the buffer
        while (getchar() != '\n');
    }

    printf("\n\n Productions are : ");
    for (i = 0; i < tot; i++)
        printf("\n\n %s -> %s", p[i].lhs, p[i].rhs);

    cnt = 0;
    sym[cnt++] = p[0].lhs[0];

    for (i = 0; i < tot; i++) {
        flag = 0;
        for (j = 0; j < strlen(sym); j++) {
            if (sym[j] == p[i].lhs[0]) {
                flag = 1;
                break;
            }
        }
        if (!flag) {
            sym[cnt++] = p[i].lhs[0];
        }
        for (k = 0; k < strlen(p[i].rhs); k++) {
            flag = 0;
            for (j = 0; j < strlen(sym); j++) {
                if (sym[j] == p[i].rhs[k]) {
                    flag = 1;
                    break;
                }
            }
            if (!flag) {
                sym[cnt++] = p[i].rhs[k];
            }
        }
    }
    printf("\n\n sym final : \n");
    for (i = 0; i < strlen(sym); i++) {
        printf("\t%c", sym[i]);
    }
    // Initialize matrices
    for (i = 0; i < 20; i++) {
        for (j = 0; j < 20; j++) {
            first[i][j] = firstp[i][j] = firsts[i][j] = last[i][j] = lastp[i][j] = lasts[i][j] = equals[i][j] = lt[i][j] = gt[i][j] = 0;
            opm[i][j] = '0';  // Initialize operator matrix
        }
    }
    return 0;
}

File: Py/P6_OPM.py
#Design a program to develop Operator Precedence Matrix (OPM).
#abc def

def display(matrix, symbols):
    # Print header row
    print("  ", " ".join(symbols))
    
    # Print matrix with corresponding symbol rows
    for i, sym in enumerate(symbols):
        print(f"{sym} ", " ".join(map(str, matrix[i])))

def main():
    productions = []
    first = [[0] * 20 for _ in range(20)]
    firstp = [[0] * 20 for _ in range(20)]
    firsts = [[0] * 20 for _ in range(20)]
    last = [[0] * 20 for _ in range(20)]
    lastp = [[0] * 20 for _ in range(20)]
    lasts = [[0] * 20 for _ in range(20)]
    equals = [[0] * 20 for _ in range(20)]
    lt = [[0] * 20 for _ in range(20)]
    gt = [[0] * 20 for _ in range(20)]
    opm = [['0'] * 20 for _ in range(20)]  # Operator precedence matrix
    symbols = []
    tot = 0
    
    ans = 'n'
    
    # Input productions
    while ans == 'n':
        print("Saail Chavan KFPMSCCS016")
        lhs = input("Enter LHS: ")
        rhs = input("Enter RHS: ")
        productions.append((lhs, rhs))
        tot += 1
        ans = input("\nContinue? (y/n): ").lower()

    # Display productions
    print("\nProductions are:")
    for lhs, rhs in productions:
        print(f"\n{lhs} -> {rhs}")

    # Extract unique symbols
    symbols.append(productions[0][0][0])
    
    for lhs, rhs in productions:
        if lhs[0] not in symbols:
            symbols.append(lhs[0])
        for char in rhs:
            if char not in symbols:
                symbols.append(char)

    # Display final symbols
    print("\n\nsym final:")
    print("\t", "\t".join(symbols))

    # Initialize matrices
    for i in range(20):
        for j in range(20):
            first[i][j] = firstp[i][j] = firsts[i][j] = last[i][j] = lastp[i][j] = lasts[i][j] = equals[i][j] = lt[i][j] = gt[i][j] = 0
            opm[i][j] = '0'  # Initialize operator precedence matrix

if __name__ == "__main__":
    main()